File: spark-kernel-dsv2/src/main/scala/io/delta/spark/dsv2/utils/SchemaUtils.scala
/*
 * Copyright (2025) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.delta.spark.dsv2.utils

import scala.collection.JavaConverters._

import io.delta.kernel.types.{ArrayType => KernelArrayType, BinaryType => KernelBinaryType, BooleanType => KernelBooleanType, ByteType => KernelByteType, DataType => KernelDataType, DateType => KernelDateType, DecimalType => KernelDecimalType, DoubleType => KernelDoubleType, FloatType => KernelFloatType, IntegerType => KernelIntegerType, LongType => KernelLongType, MapType => KernelMapType, ShortType => KernelShortType, StringType => KernelStringType, StructField => KernelStructField, StructType => KernelStructType, TimestampNTZType => KernelTimestampNTZType, TimestampType => KernelTimestampType}

import org.apache.spark.sql.types.{ArrayType => SparkArrayType, BinaryType => SparkBinaryType, BooleanType => SparkBooleanType, ByteType => SparkByteType, DataType => SparkDataType, DateType => SparkDateType, DecimalType => SparkDecimalType, DoubleType => SparkDoubleType, FloatType => SparkFloatType, IntegerType => SparkIntegerType, LongType => SparkLongType, MapType => SparkMapType, ShortType => SparkShortType, StringType => SparkStringType, StructField => SparkStructField, StructType => SparkStructType, TimestampNTZType => SparkTimestampNTZType, TimestampType => SparkTimestampType}

object SchemaUtils {

  //////////////////////
  // Kernel --> Spark //
  //////////////////////

  def convertKernelSchemaToSparkSchema(kernelSchema: KernelStructType): SparkStructType = {
    SparkStructType(kernelSchema.fields().asScala.map(convertKernelFieldToSparkField).toSeq)
  }

  def convertKernelFieldToSparkField(kernelField: KernelStructField): SparkStructField = {
    SparkStructField(
      kernelField.getName,
      convertKernelDataTypeToSparkDataType(kernelField.getDataType),
      kernelField.isNullable)
  }

  def convertKernelDataTypeToSparkDataType(kernelDataType: KernelDataType): SparkDataType = {
    kernelDataType match {
      case _: KernelStringType => SparkStringType
      case _: KernelBooleanType => SparkBooleanType
      case _: KernelIntegerType => SparkIntegerType
      case _: KernelLongType => SparkLongType
      case _: KernelBinaryType => SparkBinaryType
      case _: KernelByteType => SparkByteType
      case _: KernelDateType => SparkDateType
      case kernelDecimal: KernelDecimalType =>
        SparkDecimalType(kernelDecimal.getPrecision, kernelDecimal.getScale)
      case _: KernelDoubleType => SparkDoubleType
      case _: KernelFloatType => SparkFloatType
      case _: KernelShortType => SparkShortType
      case _: KernelTimestampType => SparkTimestampType
      case _: KernelTimestampNTZType => SparkTimestampNTZType
      case kernelArray: KernelArrayType =>
        SparkArrayType(
          convertKernelDataTypeToSparkDataType(kernelArray.getElementType),
          kernelArray.containsNull())
      case kernelMap: KernelMapType =>
        SparkMapType(
          convertKernelDataTypeToSparkDataType(kernelMap.getKeyType),
          convertKernelDataTypeToSparkDataType(kernelMap.getValueType),
          kernelMap.isValueContainsNull)
      case kernelStruct: KernelStructType => convertKernelSchemaToSparkSchema(kernelStruct)
      case x => throw new IllegalArgumentException(s"unsupported data type $x")
    }
  }

  //////////////////////
  // Spark --> Kernel //
  //////////////////////

  def convertSparkSchemaToKernelSchema(sparkSchema: SparkStructType): KernelStructType = {
    new KernelStructType(
      sparkSchema.fields
        .map(convertSparkFieldToKernelField)
        .toList
        .asJava)
  }

  def convertSparkFieldToKernelField(sparkField: SparkStructField): KernelStructField = {
    new KernelStructField(
      sparkField.name,
      convertSparkDataTypeToKernelDataType(sparkField.dataType),
      sparkField.nullable)
  }

  def convertSparkDataTypeToKernelDataType(sparkDataType: SparkDataType): KernelDataType = {
    sparkDataType match {
      case SparkStringType => KernelStringType.STRING
      case SparkBooleanType => KernelBooleanType.BOOLEAN
      case SparkIntegerType => KernelIntegerType.INTEGER
      case SparkLongType => KernelLongType.LONG
      case SparkBinaryType => KernelBinaryType.BINARY
      case SparkByteType => KernelByteType.BYTE
      case SparkDateType => KernelDateType.DATE
      case sparkDecimal: SparkDecimalType =>
        new KernelDecimalType(sparkDecimal.precision, sparkDecimal.scale)
      case SparkDoubleType => KernelDoubleType.DOUBLE
      case SparkFloatType => KernelFloatType.FLOAT
      case SparkShortType => KernelShortType.SHORT
      case SparkTimestampType => KernelTimestampType.TIMESTAMP
      case SparkTimestampNTZType => KernelTimestampNTZType.TIMESTAMP_NTZ
      case sparkArray: SparkArrayType =>
        new KernelArrayType(
          convertSparkDataTypeToKernelDataType(sparkArray.elementType),
          sparkArray.containsNull)
      case sparkMap: SparkMapType =>
        new KernelMapType(
          convertSparkDataTypeToKernelDataType(sparkMap.keyType),
          convertSparkDataTypeToKernelDataType(sparkMap.valueType),
          sparkMap.valueContainsNull)
      case sparkStruct: SparkStructType => convertSparkSchemaToKernelSchema(sparkStruct)
      case x => throw new IllegalArgumentException(s"unsupported data type $x")
    }
  }
}

File: spark-kernel-dsv2/src/main/java/io/delta/spark/dsv2/utils/SchemaUtils.java
/*
 * Copyright (2025) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.spark.dsv2.utils;

import io.delta.kernel.types.ArrayType;
import io.delta.kernel.types.BinaryType;
import io.delta.kernel.types.BooleanType;
import io.delta.kernel.types.ByteType;
import io.delta.kernel.types.DataType;
import io.delta.kernel.types.DateType;
import io.delta.kernel.types.DecimalType;
import io.delta.kernel.types.DoubleType;
import io.delta.kernel.types.FloatType;
import io.delta.kernel.types.IntegerType;
import io.delta.kernel.types.LongType;
import io.delta.kernel.types.MapType;
import io.delta.kernel.types.ShortType;
import io.delta.kernel.types.StringType;
import io.delta.kernel.types.StructField;
import io.delta.kernel.types.StructType;
import io.delta.kernel.types.TimestampNTZType;
import io.delta.kernel.types.TimestampType;
import java.util.ArrayList;
import java.util.List;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.Metadata;

public class SchemaUtils {

  //////////////////////
  // Kernel --> Spark //
  //////////////////////

  /** Converts a Delta Kernel schema to a Spark schema. */
  public static org.apache.spark.sql.types.StructType convertKernelSchemaToSparkSchema(
      StructType kernelSchema) {
    List<org.apache.spark.sql.types.StructField> fields = new ArrayList<>();

    for (StructField field : kernelSchema.fields()) {
      fields.add(
          new org.apache.spark.sql.types.StructField(
              field.getName(),
              convertKernelDataTypeToSparkDataType(field.getDataType()),
              field.isNullable(),
              Metadata.empty()));
    }

    return new org.apache.spark.sql.types.StructType(
        fields.toArray(new org.apache.spark.sql.types.StructField[0]));
  }

  /** Converts a Delta Kernel data type to a Spark data type. */
  public static org.apache.spark.sql.types.DataType convertKernelDataTypeToSparkDataType(
      DataType kernelDataType) {
    if (kernelDataType instanceof StringType) {
      return DataTypes.StringType;
    } else if (kernelDataType instanceof BooleanType) {
      return DataTypes.BooleanType;
    } else if (kernelDataType instanceof IntegerType) {
      return DataTypes.IntegerType;
    } else if (kernelDataType instanceof LongType) {
      return DataTypes.LongType;
    } else if (kernelDataType instanceof BinaryType) {
      return DataTypes.BinaryType;
    } else if (kernelDataType instanceof ByteType) {
      return DataTypes.ByteType;
    } else if (kernelDataType instanceof DateType) {
      return DataTypes.DateType;
    } else if (kernelDataType instanceof DecimalType) {
      DecimalType kernelDecimal = (DecimalType) kernelDataType;
      return DataTypes.createDecimalType(kernelDecimal.getPrecision(), kernelDecimal.getScale());
    } else if (kernelDataType instanceof DoubleType) {
      return DataTypes.DoubleType;
    } else if (kernelDataType instanceof FloatType) {
      return DataTypes.FloatType;
    } else if (kernelDataType instanceof ShortType) {
      return DataTypes.ShortType;
    } else if (kernelDataType instanceof TimestampType) {
      return DataTypes.TimestampType;
    } else if (kernelDataType instanceof TimestampNTZType) {
      return DataTypes.TimestampNTZType;
    } else if (kernelDataType instanceof ArrayType) {
      ArrayType kernelArray = (ArrayType) kernelDataType;
      return DataTypes.createArrayType(
          convertKernelDataTypeToSparkDataType(kernelArray.getElementType()),
          kernelArray.containsNull());
    } else if (kernelDataType instanceof MapType) {
      MapType kernelMap = (MapType) kernelDataType;
      return DataTypes.createMapType(
          convertKernelDataTypeToSparkDataType(kernelMap.getKeyType()),
          convertKernelDataTypeToSparkDataType(kernelMap.getValueType()),
          kernelMap.isValueContainsNull());
    } else if (kernelDataType instanceof StructType) {
      return convertKernelSchemaToSparkSchema((StructType) kernelDataType);
    } else {
      // TODO: add variant type, this requires upgrading spark version dependency to 4.0
      throw new IllegalArgumentException("unsupported data type " + kernelDataType);
    }
  }

  //////////////////////
  // Spark --> Kernel //
  //////////////////////

  /** Converts a Spark schema to a Delta Kernel schema. */
  public static StructType convertSparkSchemaToKernelSchema(
      org.apache.spark.sql.types.StructType sparkSchema) {
    List<StructField> kernelFields = new ArrayList<>();

    for (org.apache.spark.sql.types.StructField field : sparkSchema.fields()) {
      kernelFields.add(
          new StructField(
              field.name(),
              convertSparkDataTypeToKernelDataType(field.dataType()),
              field.nullable()));
    }

    return new StructType(kernelFields);
  }

  /** Converts a Spark data type to a Delta Kernel data type. */
  public static DataType convertSparkDataTypeToKernelDataType(
      org.apache.spark.sql.types.DataType sparkDataType) {
    if (sparkDataType == DataTypes.StringType) {
      return StringType.STRING;
    } else if (sparkDataType == DataTypes.BooleanType) {
      return BooleanType.BOOLEAN;
    } else if (sparkDataType == DataTypes.IntegerType) {
      return IntegerType.INTEGER;
    } else if (sparkDataType == DataTypes.LongType) {
      return LongType.LONG;
    } else if (sparkDataType == DataTypes.BinaryType) {
      return BinaryType.BINARY;
    } else if (sparkDataType == DataTypes.ByteType) {
      return ByteType.BYTE;
    } else if (sparkDataType == DataTypes.DateType) {
      return DateType.DATE;
    } else if (sparkDataType instanceof org.apache.spark.sql.types.DecimalType) {
      org.apache.spark.sql.types.DecimalType sparkDecimal =
          (org.apache.spark.sql.types.DecimalType) sparkDataType;
      return new DecimalType(sparkDecimal.precision(), sparkDecimal.scale());
    } else if (sparkDataType == DataTypes.DoubleType) {
      return DoubleType.DOUBLE;
    } else if (sparkDataType == DataTypes.FloatType) {
      return FloatType.FLOAT;
    } else if (sparkDataType == DataTypes.ShortType) {
      return ShortType.SHORT;
    } else if (sparkDataType == DataTypes.TimestampType) {
      return TimestampType.TIMESTAMP;
    } else if (sparkDataType == DataTypes.TimestampNTZType) {
      return TimestampNTZType.TIMESTAMP_NTZ;
    } else if (sparkDataType instanceof org.apache.spark.sql.types.ArrayType) {
      org.apache.spark.sql.types.ArrayType sparkArray =
          (org.apache.spark.sql.types.ArrayType) sparkDataType;
      return new ArrayType(
          convertSparkDataTypeToKernelDataType(sparkArray.elementType()),
          sparkArray.containsNull());
    } else if (sparkDataType instanceof org.apache.spark.sql.types.MapType) {
      org.apache.spark.sql.types.MapType sparkMap =
          (org.apache.spark.sql.types.MapType) sparkDataType;
      return new MapType(
          convertSparkDataTypeToKernelDataType(sparkMap.keyType()),
          convertSparkDataTypeToKernelDataType(sparkMap.valueType()),
          sparkMap.valueContainsNull());
    } else if (sparkDataType instanceof org.apache.spark.sql.types.StructType) {
      return convertSparkSchemaToKernelSchema(
          (org.apache.spark.sql.types.StructType) sparkDataType);
    } else {
      throw new IllegalArgumentException("unsupported data type " + sparkDataType);
    }
  }
}
